Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Parameters/SqlTableValuedDataParameterDefinition.cs

```diff
@@ -0,0 +1,107 @@
+{
+  public const string DistinctCollectionTableTypeNameSuffix = "_distinct";
+  public bool IsDistinct { get; }
+  public IStorageTypeInformation StorageTypeInformation { get; }
+
+  public SqlTableValuedDataParameterDefinition (IStorageTypeInformation storageTypeInformation, bool isDistinct)
+  {
+    ArgumentUtility.CheckNotNull(nameof(storageTypeInformation), storageTypeInformation);
+
+    IsDistinct = isDistinct;
+    StorageTypeInformation = storageTypeInformation;
+  }
+
+  /// <summary>
+  /// Creates an <see cref="SqlTableValuedParameterValue"/> from a <paramref name="value"/> that is a collection of items whose type is compatible with the
+  /// <see cref="StorageTypeInformation"/>'s <see cref="IStorageTypeInformation.DotNetType"/>.
+  /// </summary>
+  public object GetParameterValue (object? value)
+  {
+    var enumerable = ArgumentUtility.CheckType<IEnumerable?>(nameof(value), value);
+    if (enumerable == null)
+      return DBNull.Value;
+
+    var convertedValues = new ArrayList();
+    foreach (var item in enumerable)
+    {
+      if (item != null && !StorageTypeInformation.DotNetType.IsInstanceOfType(item))
+        throw ArgumentUtility.CreateArgumentItemTypeException(nameof(value), convertedValues.Count, StorageTypeInformation.DotNetType, item.GetType());
+
+      var convertedValue = StorageTypeInformation.ConvertToStorageType(item);
+      convertedValues.Add(convertedValue);
+    }
+
+    var tableTypeNameSuffix = IsDistinct ? DistinctCollectionTableTypeNameSuffix : string.Empty;
+    var tableTypeName = $"TVP_{StorageTypeInformation.StorageDbType}{tableTypeNameSuffix}";
+
+    var sqlDbType = GetSqlDbType(StorageTypeInformation);
+    var sqlMetaData = StorageTypeInformation.StorageTypeLength.HasValue
+        ? new SqlMetaData("Value", sqlDbType, StorageTypeInformation.StorageTypeLength.Value)
+        : new SqlMetaData("Value", sqlDbType);
+
+    var columnMetaData = new[] { sqlMetaData };
+
+    var tvpValue = new SqlTableValuedParameterValue(tableTypeName, columnMetaData);
+    foreach (var convertedValue in convertedValues)
+      tvpValue.AddRecord(convertedValue);
+
+    return tvpValue;
+  }
+
+  /// <summary>
+  /// Creates a <see cref="SqlDbType.Structured"/> <see cref="SqlParameter"/> with its <see cref="SqlParameter.TypeName"/> determined by the given
+  /// <see cref="SqlTableValuedParameterValue"/>.
+  /// </summary>
+  public IDbDataParameter CreateDataParameter (IDbCommand command, string parameterName, object parameterValue)
+  {
+    ArgumentUtility.CheckNotNull(nameof(command), command);
+    ArgumentUtility.CheckNotNullOrEmpty(nameof(parameterName), parameterName);
+    var tvpValue = ArgumentUtility.CheckNotNullAndType<SqlTableValuedParameterValue>(nameof(parameterValue), parameterValue);
+
+    var sqlParameter = (SqlParameter)command.CreateParameter();
+    sqlParameter.Value = tvpValue.Any() ? tvpValue : DBNull.Value;
```

Comment:
I'm not certain that null is correct for an empty TVP. 
We should also test for Count/Length instead of empty. depending on Any-implementation it would isntantiate an iterator isntead of first testing for IReadOnlyCollection (if there's any version that does that).

Reply:
I looked it up in the existing Acta Nova implementation - the correct value is `null`, not `DBNull.Value`, so I will change that accordingly.

Reply:
Partially. ICollection<T> and ICollection are covered in .NET 8. IReadOnlyCollection is not covered.

Reply:
> I looked it up in the existing Acta Nova implementation - the correct value is `null`, not `DBNull.Value`, so I will change that accordingly. I remember that an empty collection results in an error.
> 
> With regards to `.Any()`, I would question the value of this micro-optimization vs. the value of the statement's readability. Are you sure that preventing an iterator instantiation is worth having the code look like this: `sqlParameter.Value = tvpValue.Count == 0 ? null : tvpValue;` or `sqlParameter.Value = tvpValue.Count != 0 ? tvpValue : null;` instead of: `sqlParameter.Value = tvpValue.Any() ? tvpValue : null;`

@MichaelKetting?

Reply:
Discussed with @martin-herwich we use .NET NULL, not DBNULL for EMPTY and will have a separate task to do integrtion tests for these cases.